### PR TITLE
[C++ Worker] Support overload functions for REAY_REMOTE

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?
We should support safe c++ worker API according the [doc](https://docs.google.com/document/d/1pBRUm-w2LKXTte8uXmvQ23l7gOJV2IOwirCTPjtPGWk/edit#heading=h.86fui9aqjkh6), this pr supports overload normal tasks.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

**Comment by AmeerHajAli** [comment]
@qicosmos , what is the status of this?

File: cpp/include/ray/api/overload.h [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+UNDERLOAD(cv_cv, ref_lval)
+#undef UNDERLOAD
+
+}  // namespace detail
+
+template <typename... Ts>
+constexpr detail::underload_free<Ts...> underload_free{};
+
+template <typename... Ts>
+constexpr detail::underload<Ts...> underload{};
```
Comment (kfstorm) [inline cpp/include/ray/api/overload.h:197]: What does underload mean?

Comment (qicosmos) [reply cpp/include/ray/api/overload.h:197]: Same meaning with overload.

**Review by kfstorm — APPROVED** [review]

**Comment by kfstorm** [comment]
//python/ray/serve:test_router failure should be unrelated. Merging.

**Pull request merged**